packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go

```diff
@@ -747,138 +1051,326 @@ func (dhtPeer *DHTPeer) lookupAddressDHT(address string) (peer.ID, error) {
+func (dhtPeer *DHTPeer) handleAeaEnvelopeStream(stream network.Stream) {
+	lerror, lwarn, linfo, _ := dhtPeer.getLoggers()
 
-	err = utils.WriteBytes(stream, []byte(address))
-	if err != nil {
-		return "", errors.New("ERROR while sending address to peer:" + err.Error())
-	}
+	//linfo().Msg("Got a new aea envelope stream")
 
-	msg, err := utils.ReadString(stream)
+	buf, err := utils.ReadBytes(stream)
 	if err != nil {
-		return "", errors.New("ERROR while reading target peer id from peer:" + err.Error())
+		lerror(err).Msg("while reading envelope from stream")
+		err = stream.Reset()
+		ignore(err)
+		return
 	}
-	stream.Close()
 
-	peerid, err := peer.Decode(msg)
+	// get envelope
+	msg := &dhtnode.AcnMessage{}
+	err = proto.Unmarshal(buf, msg)
 	if err != nil {
-		return "", errors.New("CRITICAL couldn't get peer ID from message:" + err.Error())
+		lerror(err).Msg("while deserializing acn aea envelope message")
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_SERIALIZATION}
+		response := &dhtnode.AcnMessage{
+			Version: dhtnode.CurrentVersion,
+			Payload: &dhtnode.AcnMessage_Status{Status: status},
+		}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytes(stream, buf)
+		ignore(err)
+		err = stream.Close()
+		ignore(err)
+		return
 	}
 
-	return peerid, nil
-}
-
-func (dhtPeer *DHTPeer) handleAeaEnvelopeStream(stream network.Stream) {
-	lerror, lwarn, linfo, _ := dhtPeer.getLoggers()
-
-	//linfo().Msg("Got a new aea envelope stream")
+	// payload is expected to be AeaEnvelope
+	var aeaEnvelope *dhtnode.AeaEnvelope
+	switch pl := msg.Payload.(type) {
+	case *dhtnode.AcnMessage_AeaEnvelope:
+		aeaEnvelope = pl.AeaEnvelope
+	default:
+		err = errors.New("Unexpected payload")
+		lerror(err).Msg("while deserializing acn aea envelope message")
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_UNEXPECTED_PAYLOAD}
+		response := &dhtnode.AcnMessage{
+			Version: dhtnode.CurrentVersion,
+			Payload: &dhtnode.AcnMessage_Status{Status: status},
+		}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytes(stream, buf)
+		ignore(err)
+		err = stream.Close()
+		ignore(err)
+		return
+	}
 
-	envel, err := utils.ReadEnvelope(stream)
+	envel := &aea.Envelope{}
+	err = proto.Unmarshal(aeaEnvelope.Envel, envel)
 	if err != nil {
-		lerror(err).Msg("while reading envelope from stream")
-		err = stream.Reset()
+		lerror(err).Msg("while deserializing acn aea envelope message")
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_SERIALIZATION, Msgs: []string{err.Error()}}
+		response := &dhtnode.AcnMessage{
+			Version: dhtnode.CurrentVersion,
+			Payload: &dhtnode.AcnMessage_Status{Status: status},
+		}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytes(stream, buf)
+		ignore(err)
+		err = stream.Close()
+		ignore(err)
+		return
+	}
+
+	remotePubkey, err := utils.FetchAIPublicKeyFromPubKey(stream.Conn().RemotePublicKey())
```

Comment:
[STEP3.b] when a peer receives an incoming envelope it validate its source PoR and deliver it to local agent or return error